### PR TITLE
Adding initial support for D3D12.

### PR DESCRIPTION
**Pull request by dave-juicelabs**

### CODE REVIEW

File: specs/d3dcommon.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+##########################################################################/
+
+
+from .winapi import *
+
+D3D_DRIVER_TYPE = Enum('D3D_DRIVER_TYPE', [
```
Comment (jrfonseca) [inline specs/d3dcommon.py:29]: We could do `from .d3d11 import foo` for most of these to avoid duplication.

**Review by jrfonseca — COMMENTED** [review]
I didn't go through the D3D12 spec files with a fine-tooth comb, but it looks great to me overall.

What are your plans here? Do you plan to continue working on this and expect some feedback? Are you looking into some insights on what could be causing the crashes documented on the last commit? Or will you be leaving the code as is for now.

Either way, if we omit remove the debug code, I'd be happy to merge.

File: specs/d3dcommon.py (hunk 1 above)
Comment (dave-juicelabs) [reply specs/d3dcommon.py:29]: I didn't want to go changing the stuff already working while trying to get D3D12 in which is why I opted to just duplicate code in these sorts of cases.

**Comment by dave-juicelabs** [comment]
I needed the D3D12 addition for my work around VKD3D so for the time being I will be making additions and changes required for me and my work that is useful for everyone. Feedback is always welcome. I want to work within the framework you have already set up and any feedback will help.

As for help, that comment was more for my work with VKD3D. I have not actually had any real issues with apitrace and getting it to work with COD. This tool is helping me debug VKD3D.

**Comment by jrfonseca** [comment]
I was just made aware of https://blog.froggi.es/d3d12-apitrace-capture-and-replay/ and https://github.com/Joshua-Ashton/apitrace/tree/d3d12

At a glance, that integration seems more ahead.  Perhaps you can see if it also works for you too?  And I'll try to figure out how to get it merged.

**Comment by dave-juicelabs** [comment]
It is further along in the respect of supporting retrace but seems to be implementing a slightly older version of D3D12 and does not work on my workloads. I am going to pull from him and merge our work together.

**Comment by Joshua-Ashton** [comment]
> It is further along in the respect of supporting retrace

It's much further along in other aspects, such as wrapping descriptor handles, and virtual addresses. Tracking of writes to mapped memory with write watch and flushing accordingly.

There's also the hell that is deserializing/reserializing the new state objects *shudders*.

> but seems to be implementing a slightly older version of D3D12

Everything that I have defined in the spec is the stuff that's available out of the box without using the Insider SDK headers right now.
Feel free to add the missing definitions to ``specs/d3d12.py``

> does not work on my workloads.

What is being pull requested here won't work in any workload because it doesn't wrap descriptors, VAs, state objects or do any memory tracking. So I am not sure what you mean?

**Comment by dave-juicelabs** [comment]
The work in the current pull request only supports dumps which is all that I required of it until now. I am working on merging the new two repos together to try and bring Joshua-Ashton's version to the newer spec.

**Comment by jrfonseca** [comment]
> I am working on merging the new two repos together to try and bring Joshua-Ashton's version to the newer spec.

@dave-juicelabs, if you could rebase your D3D12 spec changes on top of @Joshua-Ashton 's branch that would be the easiest way forward.

**Comment by dave-juicelabs** [comment]
Got a good merge going but having issues with merging in some changes to dxgitrace.py. The changes are causing crashes or seemingly "hangs" in COD: Warzone when I merge in everything.

**Comment by Joshua-Ashton** [comment]
Yeah, WarZone is known to be problematic right now but I haven't managed to find the cause yet. It's also really annoying to test. Having another repro case that wasn't a 70+GB download and filled with anticheat and anti debugging would be great.

**Comment by dave-juicelabs** [comment]
I was able to merge your branch @Joshua-Ashton and mine and have it capturing COD with the removal of the WaitFor*Object hooks in my local code. I am currently debating approaches to getting OpenExistingHeapFromAddress which is obviously dying on the replay side.

**Comment by Joshua-Ashton** [comment]
It might probably work if you replace WaitFor*Objects hooks with Detours hooks rather than IAT hooks.

**Comment by dave-juicelabs** [comment]
Already attempted that and same result, COD will exit as soon as D3D12CreateDevice is called and the hooks instantiated.

**Comment by dave-juicelabs** [comment]
There is still some work to be done but COD is capturing and playing back but not rendering. It is also the only game that I have tested with this. @Joshua-Ashton Would you be able to test this on some of the games that you worked with before?

**Comment by jrfonseca** [comment]
> I was able to merge your branch @Joshua-Ashton and mine and have it capturing COD with the removal of the WaitFor*Object hooks in my local code.

@dave-juicelabs -- perhaps I'm not looking correctly, but AFAICT this PR and https://github.com/Joshua-Ashton/apitrace/commits/d3d12 still looks like two independent branches with independent histories.

Even if rebasing one branch on top of the other is not practice, I'd prefer at least to see a merge commit of both.  I'd rather not be in the position of having to pick one over the other.

**Comment by dave-juicelabs** [comment]
No worries. I will do a merge commit today.

**Comment by jrfonseca** [comment]
> No worries. I will do a merge commit today.

@dave-juicelabs did you get a chance to do that?

**Comment by Joshua-Ashton** [comment]
Anything happening with this?

**Comment by Joshua-Ashton** [comment]
I did some more work today on https://github.com/Joshua-Ashton/apitrace/commits/d3d12 about OpenExistingHeapFromAddress, WriteBufferImmediate and a fix for IASetVertexBuffers

**Comment by jrfonseca** [comment]
> I did some more work today on https://github.com/Joshua-Ashton/apitrace/commits/d3d12 about OpenExistingHeapFromAddress, WriteBufferImmediate and a fix for IASetVertexBuffers

I just tried to build your branch, but if fail with:

```Traceback (most recent call last):
  File "/home/jfonseca/projects/apitrace/wrappers/d2d1trace.py", line 48, in <module>
    tracer.traceApi(api)
  File "/home/jfonseca/projects/apitrace/wrappers/trace.py", line 595, in traceApi
    self.traceInterfaces(api)
  File "/home/jfonseca/projects/apitrace/wrappers/trace.py", line 818, in traceInterfaces
    self.implementWrapperInterface(iface)
  File "/home/jfonseca/projects/apitrace/wrappers/trace.py", line 894, in implementWrapperInterface
    self.implementWrapperInterfaceMethod(iface, base, method)
  File "/home/jfonseca/projects/apitrace/wrappers/trace.py", line 971, in implementWrapperInterfaceMethod
    self.implementWrapperInterfaceMethodBody(interface, base, method)
  File "/home/jfonseca/projects/apitrace/wrappers/trace.py", line 982, in implementWrapperInterfaceMethodBody
    self.implementWrapperInterfaceMethodBodyEx(self, interface, base, method, '_result')
TypeError: implementWrapperInterfaceMethodBodyEx() takes 5 positional arguments but 6 were given
```
There's one self too many on

```python
    self.implementWrapperInterfaceMethodBodyEx(self, interface, base, method, '_result')
```

So I'm confused: this can't be ready for merging if it can't build...

**Comment by jrfonseca** [comment]
@Joshua-Ashton, mingw32 build is also broken for me.  I've setup continuous integration so I don't have to fix other people's builds.  If it can't build, I'm afraid it's a non-starter for me.   You might find it easier to rebase/merge your branch into master, so it picks up the GitHub Actions integration, as it's easier to setup than AppVeyor/Travis (no work necessary.)

I'm starting to think that lumping this PR with https://github.com/Joshua-Ashton/apitrace was a mistake: I shouldn't have tried to hit two birds with one stone.

So I'm going to step back: I'll merge whichever D3D12 branch that it's clean: it builds, and no glaring mistakes upon review.  No matter how much effort and great work was put into adding D3D12 support to apitrace, if it doesn't meet this bar, then I'm afraid I can't merge, as I simply don't have the time to fill the missing gap.

**Comment by Joshua-Ashton** [comment]
My branch is definitely not clean for merging, I was posting it here as reference for @dave-juicelabs.
That branch just satisfies our needs right now in Proton for D3D12 apitracing. (works in Ghostrunner, FH4, Control, HZD, Death Stranding etc)
It's very possible that other frontends and 32-bit are broken right now there.

It is also not possible to support 32-bit APITracing for D3D12 in general due to the way the VA resolver works.

I'd be willing to do some cleaning on it to make it ready to merge (it'd be a decent amount of work but it's do-able), but I also don't know what @dave-juicelabs is working on because their thing seems entirely different... :-/

**Comment by dave-juicelabs** [comment]
I apologize, I am in the process of moving across the country so things are a bit hectic for me. I just added the merge commit that I promised so long ago. This code that I committed I have only tested on COD Warzone.

As for what I am doing, I am using this to debug my own work projects so the purpose of my changes is to ensure apitrace works with some D3D12 games which currently is only COD Warzone.

I am more than happy to work with both of you to further the progress of apitrace and get these changes merged in to the master.

**Comment by Joshua-Ashton** [comment]
@dave-juicelabs Can you email me at `joshua@froggi.es`? I'd like to invite you to where we do development and stuff.

**Comment by jrfonseca** [comment]
@Joshua-Ashton wrote:
> My branch is definitely not clean for merging, I was posting it here as reference for @dave-juicelabs.

Oh I see.  I hadn't understood that.  (I somehow got impression it was fairly complete.)

> I'd be willing to do some cleaning on it to make it ready to merge (it'd be a decent amount of work but it's do-able), but I also don't know what @dave-juicelabs is working on because their thing seems entirely different... :-/

Right.  If you could sync up to avoid duplicate work that would be for the best.

For my part there's no rush in merging.  It's mostly for the benefit/convenience of others who may be interested.  I don't want to be the blocking factor, but otherwise feel free to take the time you both need.

**Comment by jrfonseca** [comment]
One final remark: I'm happy to merge in partial improvements (of either branch.)  So instead of cleanup the whole shebang, it might be easier to just get the easier bits in, and tackle the more controversial/tricky bits later.

**Comment by okias** [comment]
@dave-juicelabs @Joshua-Ashton Hello, are you planning to continue work on this PR?

**Comment by okias** [comment]
@Joshua-Ashton Hello, as we spoke on XDC 2022, would you be able to do some cleanup on your branch and send MR? No new functionality needed, just stuff you already have :) Thanks

**Pull request closed without merging**